crates/rspack_plugin_javascript/src/dependency/new_url/mod.rs

```diff
@@ -0,0 +1,136 @@
+use rspack_core::{
+  create_javascript_visitor, runtime_globals, CodeGeneratable, CodeGeneratableContext,
+  CodeGeneratableResult, Dependency, DependencyCategory, DependencyId, DependencyType, ErrorSpan,
+  JsAstPath, ModuleDependency, ModuleIdentifier,
+};
+use swc_core::common::Spanned;
+use swc_core::ecma::utils::{member_expr, quote_ident, quote_str};
+use swc_core::ecma::{ast::*, atoms::JsWord};
+
+#[derive(Debug, Eq, Clone)]
+pub struct NewUrlDependency {
+  id: Option<DependencyId>,
+  parent_module_identifier: Option<ModuleIdentifier>,
+  request: JsWord,
+  span: Option<ErrorSpan>,
+  #[allow(unused)]
+  ast_path: JsAstPath,
+}
+
+// Do not edit this, as it is used to uniquely identify the dependency.
+impl PartialEq for NewUrlDependency {
+  fn eq(&self, other: &Self) -> bool {
+    self.parent_module_identifier == other.parent_module_identifier && self.request == other.request
+  }
+}
+
+// Do not edit this, as it is used to uniquely identify the dependency.
+impl std::hash::Hash for NewUrlDependency {
+  fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+    self.parent_module_identifier.hash(state);
+    self.request.hash(state);
+    self.category().hash(state);
+    self.dependency_type().hash(state);
+  }
+}
+
+impl NewUrlDependency {
+  pub fn new(request: JsWord, span: Option<ErrorSpan>, ast_path: JsAstPath) -> Self {
+    Self {
+      id: None,
+      parent_module_identifier: None,
+      request,
+      span,
+      ast_path,
+    }
+  }
+}
+
+impl Dependency for NewUrlDependency {
+  fn id(&self) -> Option<DependencyId> {
+    self.id
+  }
+  fn set_id(&mut self, id: Option<DependencyId>) {
+    self.id = id;
+  }
+  fn parent_module_identifier(&self) -> Option<&ModuleIdentifier> {
+    self.parent_module_identifier.as_ref()
+  }
+
+  fn set_parent_module_identifier(&mut self, module_identifier: Option<ModuleIdentifier>) {
+    self.parent_module_identifier = module_identifier;
+  }
+
+  fn category(&self) -> &DependencyCategory {
+    &DependencyCategory::CommonJS
+  }
+
+  fn dependency_type(&self) -> &DependencyType {
+    &DependencyType::CjsRequire
```

Comment:
Would you align `type` and `category` with webpack's? https://github.com/webpack/webpack/blob/f58ff9beb63bd9c8d5b56d73c55cbe5b2e8ce5c7/lib/dependencies/URLDependency.js#L46 Or there might be some gap in the future with `byDependency`, etc.

Reply:
You can add a `DependencyType::NewUrl` here https://github.com/web-infra-dev/rspack/blob/216c6bac38025e21f3e18980695e552b62540bbe/crates/rspack_core/src/dependency/mod.rs#L28 with comment `new URL()` and use `DependencyCategory::Url` directly for the `category`.